data/transactions/logic/assembler.go

```diff
@@ -1802,57 +1856,141 @@ func (ops *OpStream) typeErrorf(format string, args ...interface{}) {
-			strings.Join(instruction, " "), argcount, len(ops.known.stack))
+			strings.Join(instruction.strings(), " "), expectedCnt, actualCnt)
 	} else {
-		firstPop := true
-		for i := argcount - 1; i >= 0; i-- {
-			argType := args[i]
-			stype := ops.known.pop()
-			if firstPop {
-				firstPop = false
-				ops.trace("pops(%s", argType)
-			} else {
-				ops.trace(", %s", argType)
-			}
-			if !typecheck(argType, stype) {
+		for i := expectedCnt - 1; i >= 0; i-- {
+			expectedType, actualType := args[i], ops.known.pop()
+			if !actualType.AssignableTo(expectedType) {
 				ops.typeErrorf("%s arg %d wanted type %s got %s",
-					strings.Join(instruction, " "), i, argType, stype)
+					strings.Join(instruction.strings(), " "), i,
+					expectedType.String(), actualType.String())
 			}
 		}
-		if !firstPop {
-			ops.trace(")")
-		}
+		ops.trace("pops%s ", args.String())
 	}
 
 	if len(returns) > 0 {
 		ops.known.push(returns...)
-		ops.trace(" pushes(%s", returns[0])
-		if len(returns) > 1 {
-			for _, rt := range returns[1:] {
-				ops.trace(", %s", rt)
-			}
+		ops.trace("pushes%s ", returns.String())
+	}
+}
+
+// assembleTokens works on a single set of tokens representing
+// an op and its immediate args
+func (ops *OpStream) assembleTokens(current []token) {
+	if len(current) == 0 {
+		return
+	}
+
+	// we're about to begin processing opcodes, so settle the Version
+	if ops.Version == assemblerNoVersion {
+		ops.Version = AssemblerDefaultVersion
+	}
+	if ops.versionedPseudoOps == nil {
+		ops.versionedPseudoOps = prepareVersionedPseudoTable(ops.Version)
+	}
+
+	// Check for label
+	opstring := current[0].str
+	if opstring[len(opstring)-1] == ':' {
+		ops.createLabel(opstring[:len(opstring)-1])
+		current = current[1:]
+		if len(current) == 0 {
+			ops.trace("%3d: label only\n", ops.sourceLine)
+			return
 		}
-		ops.trace(")")
+		// it is a label but more on the same line with no semi colon
+		// e.g. `...; label_a: int 1; ...`
+		opstring = current[0].str
+	}
+
+	defer ops.trace("\n")
+
+	// Get the spec for the current op
+	spec, expandedName, ok := getSpec(ops, opstring, current[1:])
+	if !ok {
+		return
+	}
+
+	ops.trace("%3d: %s\t", ops.sourceLine, opstring)
+	ops.recordSourceLine()
+	if spec.Modes == ModeApp {
+		ops.HasStatefulOps = true
+	}
+
+	args, returns := spec.Arg.Types, spec.Return.Types
+	if spec.refine != nil {
+		nargs, nreturns, err := spec.refine(&ops.known, current[1:])
+		if err != nil {
+			ops.typeErrorf("%w", err)
+		}
+		if nargs != nil {
+			args = nargs
+		}
+		if nreturns != nil {
+			returns = nreturns
+		}
+	}
+
+	ops.trackStack(args, returns, append([]token{{str: expandedName}}, current[1:]...))
+
+	spec.asm(ops, &spec, current[1:]) //nolint:errcheck // ignore error and continue, to collect more errors
+
+	if spec.deadens() { // An unconditional branch deadens the following code
+		ops.known.deaden()
+	}
+
+	if spec.Name == "callsub" || spec.Name == "retsub" {
```

Comment:
This may be incorrect but my thinking was that after a retsub we should be back in the land of the `undead` where we can track the stack